File: DataScience/DataScience101/Part 2 - Regression/Section 2.7 - Random Forest Regression/ScriptR/My_random_forest_regression.R
# Decision Tree Regression Model
#install.packages('randomForest')
library(randomForest)

# set environment
setwd("C:/Users/nbenvenuti/Desktop/Tutorial PyR/Machine Learning/Part 2 - Regression/Section 2.7 - Random Forest Regression/")


# Importing the dataset
dataset = read.csv('./Data/Position_Salaries.csv')
dataset = dataset[2:3]


# # Splitting the dataset
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Profit, SplitRatio = 0.8)
# training_set = subset(dataset, split == T)
# test_set = subset(dataset, split == F)


# Fitting SVR to the dataset
set.seed(1234)
regressor = randomForest(x=dataset[1], y=dataset$Salary, ntree = 500)


# Prediction for 6.5 years of experience salary
y_pred=predict(regressor, data.frame(Level=6.5))



# Plotting Results Random Forest Regression
library(ggplot2)
ggplot() +
  geom_point(aes(x=dataset$Level, y=dataset$Salary), col="red") + 
  geom_line(aes(x=dataset$Level, y=predict(regressor)), col="blue")



# Plotting Results Random Forest Regression with high definition and smoother curve
x_grid=seq(min(dataset$Level), max(dataset$Level), by = 0.01)
ggplot() +
  geom_point(aes(x=dataset$Level, y=dataset$Salary), col="red") + 
  geom_line(aes(x=x_grid, y=predict(regressor, newdata = data.frame(Level = x_grid, Level2=x_grid^2, Level3=x_grid^3))), col="blue") + 
  ggtitle("Truth or Bluff (Random Forest Regression)") +
  xlab("Level") +
  ylab("Salary")

File: DataScience/DataScience101/Part 2 - Regression/Section 2.7 - Random Forest Regression/ScriptPy/My_random_forest_regression.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 27 19:16:33 2018

@author: nbenvenuti
"""

# RANDOM FOREST REGRESSION

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
os.chdir('C:\\Users\\nbenvenuti\\Desktop\\Tutorial PyR\\Machine Learning\\Part 2 - Regression\Section 2.5 - Support Vector Regression (SVR)')
    
# import dataset
dataset = pd.read_csv('./Data/Position_Salaries.csv')
x = dataset.iloc[:, 1:2].values # trick to force a Matrix with only one column
y = dataset.iloc[:, 2].values # matrix of indipendent variable


# Splitting Dataset into the training and test set
# in this case we have to few lines to perform it
"""from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(x,y, test_size=0.2, random_state=0)"""

# feature scaling
"""from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
sc_y = StandardScaler()
x = sc_x.fit_transform(x)
y = sc_y.fit_transform(y)
y = y[:,0]"""


# Fitting RF to the dataset
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(n_estimators=300, criterion='mse',                                  random_state = 0)
regressor.fit(x, y)


# Predicting New Results
# need to inverse since is a transform vallue + need to create an array of one value
y_pred=regressor.predict(6.5)


# Visualizing the Random Forest
plt.scatter(x,y,color='red')
plt.plot(x,regressor.predict(x), color='blue')
plt.title('Truth or Bluff  (Deision Tree Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()


# Visualizing the Random Forest HD
x_grid = np.arange(min(x), max(x), 0.01)
x_grid = x_grid.reshape(len(x_grid), 1)
plt.scatter(x,y,color='red')
plt.plot(x_grid,regressor.predict(x_grid), color='blue')
plt.title('Truth or Bluff  (Deision Tree Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()


